editor/src/clj/editor/code/view.clj

```diff
@@ -1025,6 +1027,240 @@
+                      :indent-type
+                      (g/set-property resource-node :modified-indent-type value)
+
+                      ;; The :lines output in the resource node is uncached.
+                      ;; It reads from disk unless a value exists for the
+                      ;; :modified-lines property. This means only modified
+                      ;; or currently open files are kept in memory.
+                      :lines
+                      (g/set-property resource-node :modified-lines value)
+
+                      ;; All other properties are set on the view node.
+                      (g/set-property view-node prop-kw value))))
+          values-by-prop-kw)))
+
+(defn- set-resource-properties
+  "Return transaction steps eduction for the editable resource node changes"
+  [resource-node values-by-prop-kw]
+  (e/mapcat
+    (fn [[prop-kw value]]
+      (case prop-kw
+        :cursor-ranges (g/set-property resource-node :cursor-ranges value)
+        :regions (g/set-property resource-node :regions (filterv #(= :regions (region->prop-kw %)) value))
+        :invalidated-row (g/update-property resource-node :invalidated-rows conj value)
+        :lines (g/set-property resource-node :modified-lines value)
+        :indent-type (g/set-property resource-node :modified-indent-type value)))
+    values-by-prop-kw))
+
+(defn set-properties!
+  "Sets values of properties that are managed by the functions in the code.data module.
+  Returns true if any property changed, false otherwise."
+  [view-node undo-grouping values-by-prop-kw]
+  (if (empty? values-by-prop-kw)
+    false
+    (do (g/transact
+          (set-properties view-node undo-grouping values-by-prop-kw))
+        true)))
+
+;; endregion
+
+;; region rename
+
+(defn- handle-rename-key-pressed [view-node text rename-cursor-range swap-state ^KeyEvent e]
+  (when (= KeyCode/ENTER (.getCode e))
+    (.consume e)
+    (g/with-auto-evaluation-context evaluation-context
+      (let [resource-node (g/node-value view-node :resource-node evaluation-context)
+            lsp (lsp/get-node-lsp (:basis evaluation-context) resource-node)]
+        (swap-state assoc :done true)
+        (lsp/rename
+          lsp
+          rename-cursor-range
+          text
+          (fn on-rename-response [resource->ascending-cursor-ranges-and-replacements]
+            (ui/run-later
+              (some->
+                (g/with-auto-evaluation-context evaluation-context
+                  (when (identical? rename-cursor-range (get-property view-node :rename-cursor-range evaluation-context))
+                    (let [resource->view (->> (get-property view-node :open-views evaluation-context)
+                                              (e/keep
+                                                (fn [[view {:keys [resource]}]]
+                                                  (when (g/node-kw-instance? (:basis evaluation-context) ::CodeEditorView view)
```

Comment:
Use `node-kw-instance?` because we define the rename popup before we define the `CodeEditorView` node, since the node refers to the popup in its output.